src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs

```diff
@@ -244,9 +249,28 @@ void Log(LogLine log)
 
             try
             {
-                await foreach (var entry in channel.GetBatchesAsync(cancellationToken: cancellationToken).ConfigureAwait(false))
+                var lineCount = 0;
+                while (await channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
                 {
-                    yield return entry;
+                    var batch = new List<LogLine>();
+                    while (channel.Reader.TryRead(out var log))
+                    {
+                        if (lineCount > backlogSnapshot.Length)
```

Comment:
We might want to capture `backlogSnapshot.Length` in a variable so `backlogSnapshot` can get GC'd earlier. As it stands, it won't be available for collection until the log stops.

Reply:
Doesn't it remain in scope due to the `Contains` call anyway?

Reply:
This code bothers me a lot 😅. The only other thing that came to mind was to populate the channel from the backlog (doing de-duping there). Also deduping based on content not id bothers me

Reply:
> The only other thing that came to mind was to populate the channel from the backlog (doing de-duping there).

Not quite following your thinking on this one. Ultimately with the current order of things we need to subscribe first before taking the snapshot, so when would populate the channel from the backlog in a way that allows to de-dupe?

> Also deduping based on content not id bothers me

What id? The line number?

Reply:
> Doesn't it remain in scope due to the `Contains` call anyway?

Yeah, I missed that. Good call.

Reply:
> This code bothers me a lot 😅. The only other thing that came to mind was to populate the channel from the backlog (doing de-duping there). Also deduping based on content not id bothers me

OK @davidfowl this is reimplemented based on our discussion now. LMK what you think.